### PR TITLE
Disable hash lookup when we have the docker digest

### PR DESCRIPTION
**Pull request by illusional**
This PR is based on a comment in the Slack, and ongoing conversation in [How to configure proxies?](https://broadworkbench.atlassian.net/browse/BA-5695) (which stems from #5006).

Essentially, I don't want Cromwell to lookup my container when I provide it as a digest, because if the request fails (because Cromwell isn't respecting the system proxy), I still want call-caching to work. The main reason Cromwell seems to want to lookup the external registry is for the docker size which gets logged to metadata and doesn't seem to be used again.

Disabling the digest-lookup doesn't actually stop the digest being looked up and disabled call-caching anyway.

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/preparation/JobPreparationActor.scala [hunk 1]
```diff
@@ -149,9 +149,9 @@ class JobPreparationActor(workflowDescriptor: EngineWorkflowDescriptor,
       case Success(_: DockerImageIdentifierWithoutHash) if !hasDockerDefinition => 
         lookupKvsOrBuildDescriptorAndStop(inputs, attributes, NoDocker, None)
 
-      // Even if the docker image has a hash, we need to (try to) find out the size, so send a request
-      case Success(dockerImageId: DockerImageIdentifierWithHash) =>
-        sendDockerRequest(dockerImageId)
+      // If there's a digest, we'll skip looking up the container because it just looks up the compressed size
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/preparation/JobPreparationActor.scala:152]: Even if a docker image hash is provided, Cromwell issues the Docker request to get the docker image size for [adjusting boot disk size](https://github.com/broadinstitute/cromwell/blob/develop/supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala#L185) on the Google backend. So this behavior could perhaps be a configuration opt-in but I don't think it should be the default.

**Comment by illusional** [comment]
Thanks @mcovarr, I've moved this back to draft state, as it's something I'll need to put behind a configuration option, so requires more work.

**Comment by rhpvorderman** [comment]
@illusional Thanks for this. I have been looking into (but not having time for) an easy option that would disable the hash lookup altogether. Cromwell connecting to quay.io while quay.io is down causes crashes we do not want in production. There is a configuration option for this. So it was easy.

Unfortunately the hash lookup is coupled with the call-caching mechanic. No hash, no cache. Which is something to be aware of.

I was wondering if the easiest way wouldn't be to have the lookup be a command in the config. Just like `docker_kill` there could be a `docker_lookup_hash`. That way you can override the default with a custom command that returns a string (https://stackoverflow.com/a/39376254). 

For example:
```
$ docker inspect --format='{{index .RepoDigests 0}}' mysql:5.6
mysql@sha256:19a164794d3cef15c9ac44754604fa079adb448f82d40e4b8be8381148c785fa
```
This does NOT need the internet.

Similarly, this would enable hash-lookup for singularity users as well without internet.

**Comment by illusional** [comment]
@rhpvorderman ^^ would just need to keep in mind they do the lookup to get the docker size.

After BCC2020 I want to revisit this PR, and could allow you to turn off the digest but keep call caching on?

The only catch i imagine is that you'd get a cache miss if you ran with / without that flag (as a digest vs floating tag would miss).

I don't know for sure other technical challenges apart from that.

**Comment by rhpvorderman** [comment]
>  would just need to keep in mind they do the lookup to get the docker size.

@illusional well I suppose a `docker_size` argument can just as easily be implemented. After pulling the image can be queried for size.

> After BCC2020 I want to revisit this PR, and could allow you to turn off the digest but keep call caching on?

Nope. Not yet anyway. And the code is intrically linked, so it is not going to be a one-liner fix. So this is why I have postponed working on this.

This is an interesting thing to revisit at a later date. We use singularity containers on a SLURM backend, using the `singularity-permanent-cache` program to pull the images. For us that really works well, and our login node has contact with the internet, so this change is not really urgent. But for stability it is always nice if an internet connection is not required anymore after all the images are there.

**Comment by illusional** [comment]
@rhpvorderman, have you considered tagging your WDL tasks with the tag AND digest (or just digest)?
You get the benefit of no internet response, still get your permanent docker containers and you could ensure your tasks don't change if a new floating tag was pushed - promoting better reproducibility.

**Comment by rhpvorderman** [comment]
@illusional All our tasks work with [biocontainers](https://biocontainers.pro/). These do not change their tags. It works with bioconda and conda in the background. Tags match bioconda builds see [here for an example](quay.io/biocontainers/gatk4).

We choose tags because it is really easy for the user to see which versions of the tools are used (like [here](https://github.com/biowdl/germline-DNA/blob/develop/dockerImages.yml)).

I did not find any information about using tag AND digest in the [docker documentation](https://success.docker.com/article/images-tagging-vs-digests). How is that possible :confused:?

**Comment by illusional** [comment]
Honestly I don't know if it's valid, I just know that the Cromwell container regex will try to pattern match for it:

https://github.com/broadinstitute/cromwell/blob/develop/dockerHashing/src/main/scala/cromwell/docker/DockerImageIdentifier.scala

**Comment by illusional** [comment]
Closing in favour of https://github.com/broadinstitute/cromwell/pull/6140

**Pull request closed without merging**